cmd/fs_handlers.go

```diff
@@ -77,6 +78,39 @@ func handleStageDirectory(ctx *cli.Context, ctl *client.Client, root, repoRoot s
 	err := filepath.Walk(root, func(childPath string, info os.FileInfo, err error) error {
 		repoPath := filepath.Join("/", repoRoot, childPath[len(root):])
 
+		if info.Mode() & os.ModeSymlink != 0 {
+			resolvedPath, err := filepath.EvalSymlinks(childPath)
+			if err != nil {
+				if os.IsNotExist(err) {
+					log.Warningf("%v symlink points to non existing file or directory", childPath)
+					return nil
+				}
+				if err.Error() == "EvalSymlinks: too many links" {
```

Comment:
Can't we just return the error instead of remapping the error to a log message that kinda says the same as the error? I mean for the logic it shouldn't matter if the file does not exist, is a cycle or failed for e.g. permission issues.

Reply:
If we return a error, it will interrupt the rest of the staging. With a warning, we let user know that the request is impossible for only this link.

I tried both ways. It presents less surprises to a user if we batch stage as much as possible, and skip things which `brig` does not support vs stopping at the first filesystem incompatible link.

Would you agree?

Another option is to process errors in the caller and decide which is critical and which is not there. This way, we can return a non zero exit code, so a user can do some `cli` logic around to decide if it is mission critical or not.

I really need a second opinion here.

Reply:
> Would you agree?

Hmm, don't fully agree. Not so much about whether traversal should be stopped or not (that might be preference thing), but rather how the current error handling is implemented:

* You ignore the error when symlink is broken or results in a loop.
* You error out if the symlink could not be read for other reasons (permissions etc.)

I don't see a clear line between either cases, therefore I'd vote to handle them equally. In both cases it results in at least one file that could not be added. Handling equally would also rid you from having to add branches for specific error causes.

About whether to stop or not: Maybe that could be an option too (``--ignore-errors``) which would apply to regular files and directories too (if they could not be read due to whatever reasons). It would not be the default, so we would error out to inform users that something is wrong. My thinking for the option is that there might be different use cases:

* "Live" usage where a user reads the log messages. Users might wish to ignore errors if they deem it safe.
* "Automated" usage by a script that just fires the command. Errors should normally lead to a script abort here.

> Another option is to process errors in the caller and decide which is critical and which is not there. This way, we can return a non zero exit code, so a user can do some cli logic around to decide if it is mission critical or not.

Would probably require interactivity (i.e. ask on the cmdline), which I'm not a big fan of in cli tools.

Reply:
You logic is more consistent. Error is a error. I will modify accordingly.